[fxcop] Wox.Infrastructure

avneet-kr opened the pull request:
## Summary of the Pull Request

_What is this about?_
This PR enables FxCop for Wox.Infrastructure and contains all fixes for Fxcop errrors

## PR Checklist
* [x] Applies to #5129 
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_

1. Enabled FxCopAnalyzer for Wox.Infrastructure
2. CA5364: Do not use deprecated security protocols (Remove unused code)
3. CA1724: Type names should not match namespaces (Renamed Http.cs to HttpClient.cs, Renamed Plugin.cs to RunPlugin.cs, renamed Settings.cs to PowerToysRunSettings.cs)
4. CA1033: Interface methods should be callable by child types (Added sealed modifier to class)
5. CA5350: Do Not Use Weak Cryptographic Algorithms (Suppressed)
6. CA2007: Do not directly await a Task (Suppressed)
7. CA1716: Identifiers should not match keywords
8. CA1044: Properties should not be write only
9. CA2234: Pass System.Uri objects instead of strings
10. CA2000: Dispose objects before losing scope 
11. CA1308: Normalize strings to uppercase
12. CA1308: Normalize strings to uppercase
13. CA1060: Move P/Invokes to NativeMethods class
14. CA1031: Do not catch general exception types
15. CA1054: URI parameters should not be strings
16. CA2227: Collection properties should be read only
17. CA1822: Mark members as static
18. CA1805: Do not initialize unnecessarily.
19. CA1707: Identifiers should not contain underscores
20. CA1820: Test for empty strings using string length
21. CA1802: Use Literals Where Appropriate
22. CA1304: Specify CultureInfo / CA1305: Specify IFormatProvider / CA1307: Specify StringComparison for clarity
23. CA1062: Validate arguments of public methods
24. CA1041: Provide ObsoleteAttribute message
25. CA1052: Static holder types should be Static or NotInheritable

## Validation Steps Performed

_How does someone test & validate?_
All changes were validated by running unit tests and building PowerToys.sln, PowerToysSetup.sln and PowerToysBootstrapper.sln from a clean state

avneet-kr committed:
Merge pull request #1 from microsoft/master

Pulling in changes from microsoft/PowerToys

avneet-kr committed:
Merge pull request #2 from microsoft/master

Merging in changes from microsoft/master

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
CA1052: Static holder types should be Static or NotInheritable

avneet-kr committed:
CA1041: Provide ObsoleteAttribute message

avneet-kr committed:
CA1062: Validate arguments of public methods

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
CA1304: Specify CultureInfo / CA1305: Specify IFormatProvider / CA1307: Specify StringComparison for clarity

avneet-kr committed:
CA1802: Use Literals Where Appropriate

avneet-kr committed:
CA1820: Test for empty strings using string length

avneet-kr committed:
CA1707: Identifiers should not contain underscores

avneet-kr committed:
CA1805: Do not initialize unnecessarily.

avneet-kr committed:
CA1822: Mark members as static

avneet-kr committed:
CA2227: Collection properties should be read only

avneet-kr committed:
CA1054: URI parameters should not be strings

avneet-kr committed:
CA1031: Do not catch general exception types

avneet-kr committed:
CA1060: Move P/Invokes to NativeMethods class

avneet-kr committed:
CA1308: Normalize strings to uppercase

avneet-kr committed:
CA2000: Dispose objects before losing scope / CA2234: Pass System.Uri objects instead of strings

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
CA2234: Pass System.Uri objects instead of strings

avneet-kr committed:
CA1044: Properties should not be write only

avneet-kr committed:
CA1716: Identifiers should not match keywords

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
CA2007: Do not directly await a Task

avneet-kr committed:
CA5350: Do Not Use Weak Cryptographic Algorithms (Suppressed)

avneet-kr committed:
CA1724: Type names should not match namespaces (renamed Settings.cs to PowerToysRunSettings.cs)

avneet-kr committed:
CA1033: Interface methods should be callable by child types (Added sealed modifier to class)

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
CA2007: Do not directly await a Task (Suppressed)

avneet-kr committed:
CA1724: Type names should not match namespaces (Renamed Plugin.cs to RunPlugin.cs)

avneet-kr committed:
CA1724: Type names should not match namespaces (Renamed Http.cs to HttpClient.cs)

avneet-kr committed:
Merge branch 'master' of github.com:avneet-kr/PowerToys into master

avneet-kr committed:
CA5364: Do not use deprecated security protocols (Remove unused code)

avneet-kr committed:
Enabled FxCopAnalyzer for Wox.Infrastructure

avneet-kr committed:
Merge branch 'master' into FxCopAnalyzer_WoxInfrastructure

avneet-kr committed:
fixed comment

alekhyareddy28 commented on src/modules/launcher/Wox.Infrastructure/UserSettings/PowerToysRunSettings.cs:
```diff
@@ -126,12 +126,12 @@ public int MaxResultsToShow
 
-        [Obsolete]
+        [ObsoleteAttribute("This method is obsolete.")]
         public double Opacity { get; set; } = 1;
 
-        [Obsolete]
+        [ObsoleteAttribute("This method is obsolete.")]
         public OpacityMode OpacityMode { get; set; } = OpacityMode.Normal;
```
If it's obsolete and not used, we can go ahead and remove it.

alekhyareddy28 commented on src/modules/launcher/Wox.Infrastructure/Storage/JsonStorage`1.cs:
```diff
@@ -106,7 +106,7 @@ private void LoadDefault()
 
         private void BackupOriginFile()
         {
-            var timestamp = DateTime.Now.ToString("yyyy-MM-dd-HH-mm-ss-fffffff", CultureInfo.CurrentUICulture);
```
Any reason for changing this? I don't think we should change it to maintain backward compatibility.

avneet-kr replied:
Fxcop complains about CurrentUICulture

`CA1305	'JsonStorage<T>.BackupOriginFile()' passes 'CultureInfo.CurrentUICulture' as the 'IFormatProvider' parameter to 'DateTime.ToString(string, IFormatProvider)'. This property returns a culture that is inappropriate for formatting methods.`

alekhyareddy28 replied:
Sounds good. Taking a deeper look at the code it seems like that value `timestamp` is being used to append to a corrupt json file. It shouldn't matter which culture we use.

alekhyareddy28 commented on src/modules/launcher/Wox.Infrastructure/Hotkey/HotkeyModel.cs:
```diff
@@ -82,7 +82,7 @@ private void Parse(string hotkeyString)
             }
 
-            List<string> keys = hotkeyString.Replace(" ", string.Empty).Split('+').ToList();
+            List<string> keys = hotkeyString.Replace(" ", string.Empty, StringComparison.Ordinal).Split('+').ToList();
```
I think these are related to key codes so maybe it's better to use invarient here?

alekhyareddy28 commented on src/modules/launcher/Wox.Infrastructure/Http/HttpClient.cs:
```diff
@@ -53,16 +45,28 @@ public static IWebProxy WebProxy()
         }
 
-        public static async Task<string> Get([NotNull] string url, string encoding = "UTF-8")
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "CA2007:Consider calling ConfigureAwait on the awaited task", Justification = "Warning is intended for libraries, not application code")]
```
Why did ConfigureAwait(false) not work in this case?

avneet-kr commented on src/modules/launcher/Wox.Infrastructure/Image/ImageHashGenerator.cs:
```diff
@@ -38,8 +42,9 @@ public string GetHashFromImage(ImageSource imageSource)
                 }
             }
-            catch
+            catch (System.Exception e)
```
Unfortunately, using `Exception` instead of `System.Exception` gives me a compiler error since it clashes with `Wox.Infrastructure.Exception`
`CS0118	'Exception' is a namespace but is used like a type	`

alekhyareddy28 replied:
Sounds good. thanks!

avneet-kr committed:
Addressed comments

- Changed Ordinal to InvariantCulture
- Added comments
- Removed unused obsolete code
- Removed unused method (CA2007: Do not directly await a Task)

avneet-kr commented:
Updated PR to address comments- 
- Changed Ordinal to InvariantCulture
- Added comments
- Removed unused obsolete code
- Removed unused method (CA2007: Do not directly await a Task)

ryanbodrug-microsoft commented on src/modules/launcher/Wox.Infrastructure/Helper.cs:
```diff
@@ -80,6 +80,7 @@ public static string Formatted<T>(this T t)
         }
 
         // Function to run as admin for context menu items
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "Exception is being logged")]
```
@avneet-kr As discussed offline, we don't want to set a standard for ignoring this error by default.  Perhaps change the justification method to something like "Justification=Suppressing this to enable FxCop.  We are logging the exception, and going forward general exceptions should not be caught"

alekhyareddy28 reviewed (approved):
Tested it out. LGTM! Please feel free to merge it after updating the FxCop justification as ryan suggestion.

avneet-kr committed:
Addressed comments - fixed justification for CA1031 suppression

avneet-kr committed:
Addressed comments - Fixed justification for CA1031 suppression in Wox.Core/Wox.Plugin

avneet-kr commented:
Updated PR - Fixed Justification for all suppressed CA1031, including for Wox.Core and Wox.Plugin
